### DISCUSSION TITLE
Translation with language file

### DISCUSSION THREAD

**Discussion by bestmomo** [filamentphp/filament#2103] [Help]
### Package

filament/filament

### Package Version

v2.0

### Laravel Version

v9.2

### Livewire Version

v2.6

### PHP Version

PHP 8.1.4

### Bug description

I try to translate a resource name using a language files (**lang/fr.json**) but I always get the generated name.

### Steps to reproduce

I created the resource **Author** and in admin menu and list title **Authors** is generated, I just add:
{`    
   "Authors": "Auteurs"
}`
in my **lang/fr.json**. But translation doesn't work.

**Comment by danharrin** [comment]
You need to override the specific methods on the resource class to return the translation string of your choice. For example, `getPluralLabel()`.

**Comment by bestmomo** [comment]
Hello,

Thanks for answer but if it is necessary to overload all the methods to obtain the translations, it is because the package was not designed for internationalization.

**Reply by danharrin** [reply]
No. The package was designed for internationalisation from the start.

Using localisation keys that map to English strings like that is actually poor Laravel localisation practice, especially for large projects. You should instead create keys that are redundant of the English version of the string, in a nested structure in your application.

In many languages, translations can be different depending on app context. Simply translating one string to another in all places is not appropriate sometimes. And automatically translating strings in the UI is problematic for some users, as there is no way to then disable this functionality.

Our current implementation is the most flexible for all users.

**Reply by bestmomo** [reply]
Thanks, I get the point now.

Would be nice to add a chapter in documentation on this point.

**Reply by Zacktamondo** [reply]
@danharrin pardon me asking but what is the ideal way to work with translation in filament, I searched a lot and didn't find anything about the subject.
If you can provide me with an example that would be great.
I tried like the author of the issue @bestmomo  mentioned and what you offered as a solution, but the problem is when using the `__` helper function it doesn't work and the titles appears as the key and nothing else.
Thank you for your time and help in advance.

**Reply by danharrin** [reply]
Can you send the code you tried and the file you tried it in?

**Reply by Zacktamondo** [reply]
Sure, here it is:
```
    public static function getPluralModelLabel(): string
    {
        return   __('lang.admin.about_us');
    }
}

```
and this is how I put it :
```
lang
├── ar
│   └── admin.php
└── en
    ├── admin.php
```
admin ar
```
<?php

return [
    'about_us' => 'About us in Ar',
];
```
admin en
```
<?php

return [
    'about_us' => 'About Us',
];

```
I tried putting it inside `resources/lang` but same issue.
The output is:
`Lang.admin.about Us`

**Reply by danharrin** [reply]
Remove lang. from the key

**Reply by danharrin** [reply]
Remove lang. from the key

**Reply by Zacktamondo** [reply]
The output become:
`Admin.about Us`

**Reply by Zacktamondo** [reply]
So looks like it doesn't work when I put inside the `lang` folder as in Laravel 9 and up.
I must put it inside the `resources/lang`.
Is there anything missing on my end or filament doesn't support `lang` in the root of the project?

**Reply by danharrin** [reply]
It supports both directories

**Reply by Zacktamondo** [reply]
What can be the problem, in your opinion?

**Reply by danharrin** [reply]
I don't know, this is a basic Laravel concept though and I doubt its anything to do with Filament

**Reply by ihahachi** [reply]
use this function and add __
` public static function getPluralLabel(): string
    {
        return __('Auteurs');
    }`

**Comment by joligoms** [comment]
Not too proud of what I did, but it seems to be working properly...

I created a translation file at ./lang/vendor/filament/[language]/resources/resource-names.php with the following structure to be defined:

```php
<?php

return [
    'article' => [
        'singular' => 'artigo',
    ],
    'user' => [
        'singular' => 'utilizador',
        'plural' => 'utilizadores',
    ],
];
```

And then, I have extended the Filament\Resources\Resource class to the App\Filament\Resource class with the following code:

```php
<?php

namespace App\Filament\Resources;

use Filament\Resources\Resource as FilamentResource;
use Illuminate\Translation\MessageSelector;
use Lang;
use Str;

abstract class Resource extends FilamentResource {
    public static function getModelLabel(): string
    {
        $modelLabel = static::getModelLabelFromClass(static::getModel());

        $translationKey = "filament::resources/resource-names.$modelLabel.singular";

        return Lang::has($translationKey) 
            ? __($translationKey) 
            : $modelLabel;
    }

    public static function getPluralModelLabel(): string
    {
        if (filled($label = static::$pluralModelLabel ?? static::getPluralLabel())) {
            return $label;
        }

        $originalModelLabel = static::getModelLabelFromClass(static::getModel());
        $pluralLabelKey = "filament::resources/resource-names.$originalModelLabel.plural";

        if (Lang::has($pluralLabelKey)) {
            return __($pluralLabelKey);
        }

        if (static::localeHasPluralization()) {
            return Str::plural(static::getModelLabel());
        }

        return static::getModelLabel();
    }

    protected static function getModelLabelFromClass(string $model): string
    {
        return (string) Str::of(class_basename($model))
            ->kebab()
            ->replace('-', ' ');
    }

    protected static function localeHasPluralization(): bool
    {
        return (new MessageSelector())->getPluralIndex(app()->getLocale(), 10) > 0;
    }
}
```
And then I extend this custom Resource class for ArticleResource, UserResource and so on...


Sorry for bleeding your eyes, but please tell me if I could improve anything to make it cleaner.

**Reply by alisalehi1380** [reply]
tnks @joligoms for your solution. 👌
IMO, based on `getModelLabel()` and `getPluralModelLabel()` in `Filament\Resources\Resource` these codes are more correct and cleaner! 😊

```php
<?php

namespace App\Filament\Resources;

use Filament\Resources\Resource as FilamentResource;
use Illuminate\Support\Facades\Lang;
use Illuminate\Support\Str;
use function Filament\Support\get_model_label;
use function Filament\Support\locale_has_pluralization;

abstract class Resource extends FilamentResource {
    public static function getModelLabel(): string
    {
        $modelLabel = get_model_label(static::getModel());

        $translationKey = "filament::resources/resource-names.$modelLabel.singular";

        return Lang::has($translationKey)
            ? __($translationKey)
            : static::$modelLabel ?? static::getLabel() ?? $modelLabel;
    }

    public static function getPluralModelLabel(): string
    {
        if (filled($label = static::$pluralModelLabel ?? static::getPluralLabel())) {
            return $label;
        }

        //Start of additions
        $originalModelLabel = get_model_label(static::getModel());
        $pluralLabelKey = "filament::resources/resource-names.$originalModelLabel.plural";

        if (Lang::has($pluralLabelKey)) {
            return __($pluralLabelKey);
        }
        //End of additions

        if (locale_has_pluralization()) {
            return Str::plural(static::getModelLabel());
        }

        return static::getModelLabel();
    }
}
```

**Comment by AHS12** [comment]
I am using this pattern. so far works great. We can override these

 2 method in our resource file
   
```php
      public static function getModelLabel(): string
      {
          return __('resources.role');
      }
      public static function getPluralLabel(): ?string
      {
          return __('resources.role.pural');
      }
```
  and for custom pages you can use this method
 ```php
    public static function getNavigationLabel(): string
    {
        return __('pages.profile.settings');
    }
```

   ![Screenshot 2023-10-12 213850](https://github.com/filamentphp/filament/assets/25058208/56b97d31-cb42-4be5-b57f-7570362209dc)
